environment.go

```diff
@@ -230,19 +235,38 @@ func (env environment) runCommand(ctx context.Context, cmd *exec.Cmd, timeout ti
 		// to the child process, so wait for it to die
 		select {
 		case <-time.After(15 * time.Second):
-			cmd.Process.Kill()
+			_ = cmd.Process.Kill()
 		case <-cmdErrChan:
 		}
 		return ctx.Err()
 	}
 }
 
-func (env environment) execGoGet(ctx context.Context, modulePath, moduleVersion string) error {
+// execGoGet runs "go get -d -v" with the given module/version as an argument.
+// Also allows passing in a second module/version pair, meant to be the main
+// Caddy module/version we're building against; this will prevent the
+// plugin module from causing the Caddy version to upgrade, if the plugin
+// version requires a newer version of Caddy.
+func (env environment) execGoGet(ctx context.Context, modulePath, moduleVersion, caddyModulePath, caddyVersion string) error {
 	mod := modulePath
 	if moduleVersion != "" {
 		mod += "@" + moduleVersion
 	}
-	cmd := env.newCommand("go", "get", "-d", "-v", mod)
+	caddy := caddyModulePath
+	if caddyVersion != "" {
+		caddy += "@" + caddyVersion
+	}
+
+	// for some reason, using an empty string as an additional
+	// argument to "go get" breaks the command, so we're using
+	// an if statement to avoid it.
+	var cmd *exec.Cmd
+	if caddy != "" {
+		cmd = env.newCommand("go", "get", "-d", "-v", mod, caddy)
+	} else {
+		cmd = env.newCommand("go", "get", "-d", "-v", mod)
+	}
```

Comment:
I thought this was completely baffling, but it was the only way I could get it to work. If I ran `env.newCommand("go", "get", "-d", "-v", mod, "")` it would just totally break. Wouldn't mind a second pair of eyes on this so I can give a better explanation in the comment.

Reply:
I tried faking the issue to replicate it by having `cmd := env.newCommand("go", "get", "-d", "-v", mod, "")`, but it still worked fine. What do you mean by break? What did you see? I'm actually failing to see a scenario where the variable `caddy` is an empty string 😅  When does that happen?

Reply:
With this change applied to the branch:

```diff
diff --git a/environment.go b/environment.go
index 961d888..e82be4a 100644
--- a/environment.go
+++ b/environment.go
@@ -257,16 +257,7 @@ func (env environment) execGoGet(ctx context.Context, modulePath, moduleVersion,
                caddy += "@" + caddyVersion
        }
 
-       // for some reason, using an empty string as an additional
-       // argument to "go get" breaks the command, so we're using
-       // an if statement to avoid it.
-       var cmd *exec.Cmd
-       if caddy != "" {
-               cmd = env.newCommand("go", "get", "-d", "-v", mod, caddy)
-       } else {
-               cmd = env.newCommand("go", "get", "-d", "-v", mod)
-       }
-
+       cmd := env.newCommand("go", "get", "-d", "-v", mod, caddy)
        return env.runCommand(ctx, cmd, env.timeoutGoGet)
 }

```

Notice the extra space in between

```
$ go build && ./xcaddy build v2.3.0 --with github.com/caddy-dns/cloudflare
2022/04/06 16:31:23 [INFO] Temporary folder: /tmp/buildenv_2022-04-06-1631.4279575882
2022/04/06 16:31:23 [INFO] Writing main module: /tmp/buildenv_2022-04-06-1631.4279575882/main.go
2022/04/06 16:31:23 [INFO] Initializing Go module
2022/04/06 16:31:23 [INFO] exec (timeout=10s): go mod init caddy 
go: creating new go.mod: module caddy
go: to add module requirements and sums:
	go mod tidy
2022/04/06 16:31:23 [INFO] Pinning versions
2022/04/06 16:31:23 [INFO] exec (timeout=0s): go get -d -v github.com/caddyserver/caddy/v2@v2.3.0  
caddy imports
	github.com/caddy-dns/cloudflare: cannot find module providing package github.com/caddy-dns/cloudflare
2022/04/06 16:31:24 [FATAL] exit status 1
```

I have _no idea_ why this happens. Big WAT

Reply:
After some experimentation (here: https://go.dev/play/p/O8J21LNQr6N), we find this reasonable. The `exec.Command` hands over the emptiness as input (argument) to the program. Removing the `""` rersults in listing the dirs of `/`. Running `ls -l ""` locally gives exactly the same results. Basically this is down to question of semantics -- do we know if the command we're running accepts empty strings as reasonable and acceptable argument or not?

Reply:
I adjusted the comment :+1: thanks for confirming! :tada: